react/AppIcon/index.jsx

```diff
@@ -0,0 +1,88 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import classNames from 'classnames'
+import styles from './styles.styl'
+
+import Icon from '../Icon'
+
+import appDefaultIcon from '../../assets/icons/ui/icon-cube.svg'
+
+export class AppIcon extends Component {
+  constructor(props, context) {
+    super(props, context)
+    this.state = {
+      status: 'idle'
+    }
+  }
+
+  async componentDidMount() {
+    const { app } = this.props
+    const { client } = this.context
+    const cozyClient = client && client.link && client.link.client
+    if (!cozyClient || typeof cozyClient.fetch !== 'function') {
+      console.warn(
+        'AppIcon needs CozyClient and to be called into a <CozyClientProvider /> component.'
+      )
+      this.setState({ status: 'failed' })
+      return
+    }
+    const icon = await this.fetchIcon(app.links.icon, cozyClient)
+    this.setState({ icon, status: icon ? 'done' : 'failed' })
+  }
+
+  async fetchIcon(url, cozyClient) {
+    this.setState({ status: 'fetching' })
+    let icon
+    try {
+      const resp = await cozyClient.fetch('GET', url)
+      if (!resp.ok)
+        throw new Error(`Error while fetching icon ${resp.statusText}: ${url}`)
+      icon = await resp.blob()
+    } catch (error) {
+      console.warn(error.message)
+      return
+    }
+    // check if MIME type is an image
+    if (!icon.type.match(/^image\/.*$/)) {
+      console.warn(`App icon ${url} is not an image.`)
+      return
+    }
+    return URL.createObjectURL(icon)
```

Comment:
IMHO all the functionality that is not connected to the component should be extracted in a helper method.

```
async fetchIcon() {
  this.setState({ status: 'fetching' })
  try {
      this.props.fetchIcon(this.props.app)
      this.setState({ status: 'done' })
  } catch (e) {
      this.setState({ status: 'error' })
  }
}
```

This would let the easier change of implementation of fetchIcon.

Reply:
I don't see the benefits here. Why extracting this method now ? If `this.props.fetchIcon` is already injected we can change its implementation.

Reply:
Yes, "helper" was perhaps not the right word, if it is extracted from the component, that's alright ;)